process: add source-map support to stack traces

bcoe opened the pull request:
This uses [SourceMap.js](https://chromium.googlesource.com/v8/v8/+/3.30.3/tools/SourceMap.js?autodive=0%2F) (shipped with V8) and the cache developed in https://github.com/nodejs/node/pull/28960, to add source-map support to stack traces.

**How it works:**

If you run your program with the flag `--source-map-support`, and a source-map is found and cached for a file, stack traces are annotated with both the original and transpiled source positions:

```
Error: goodbye
    at  /Users/bencoe/oss/node-1/test/fixtures/source-map/inline-throw.min.js:1:43
         -> /Users/bencoe/oss/node-1/test/fixtures/source-map/throw.js:8:2
    at  Immediate.<anonymous> (/Users/bencoe/oss/node-1/test/fixtures/source-map/inline-throw.min.js:1:60)
         -> /Users/bencoe/oss/node-1/test/fixtures/source-map/throw.js:8:2
    at  processImmediate (internal/timers.js:439:21)
```

Providing both the original and transpiled source if possible.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)
- [x] test with source-map generated by uglify.
- [x] test with source-map generated by babel.
- [x] test with source-map generated by TypeScript.
- [x] test with source-map generated by nyc/Istanbul.

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

This grows out of continuing conversation [here](https://github.com/nodejs/tooling/issues/40).

~And relies on https://github.com/nodejs/node/pull/28960, which I would ideally like to land first.~ (landed).

benjamingr commented on doc/api/cli.md:
```diff
@@ -1126,9 +1126,19 @@ variable is strongly discouraged.
+[Source Map][] data to the directory provided as an argument (coverage
+information is written as JSON to files with a `coverage` prefix).
+
+`NODE_V8_COVERAGE` will automatically propagate to subprocesses, making it
```
```suggestion
`NODE_V8_COVERAGE` will automatically propagate to subprocesses and worker threads, making it
```

?

benjamingr commented on lib/internal/errors.js:
```diff
@@ -40,9 +40,42 @@ const prepareStackTrace = (globalThis, error, trace) => {
-  return `${errorString}\n    at ${trace.join('\n    at ')}`;
+
+  const { getOptionValue } = require('internal/options');
+  if (getOptionValue('--experimental-source-maps')) {
```
I am for caching `getOptionValue` but I am not sure I agree this is a hot path. Errors are an exceptional path and we have always treated them as something relatively rare where we want to optimize for user experience rather than speed.

devsnek replied:
well as long as we can make them fast, i see no reason not to :smile:

bcoe replied:
I'm more e worried about the check in maybeCacheSoureMap, which is a hot path; I'll cache the value in both places 👌

benjamingr commented on test/fixtures/source-map/sigint.js:
```diff
@@ -0,0 +1,8 @@
+  const c = 102;
+}
+process.kill(process.pid, "SIGINT");
+//# sourceMappingURL=https://http.cat/402
```
I love these - but I would still prefer it if we did not have domains we do not control here since this can turn into a malicious domain at some point.

bcoe replied:
fair point, just use something we control like nodejs.org, for test?

benjamingr replied:
Yeah

benjamingr commented:
Code LGTM with devsnek's comments addressed. I'll also play with this locally a bit and see if it behaves consistently. Thanks for this this is a big usability improvement in core for a lot of our users!

(Note this will probably have to be semver-major and with a CITGM run because it changes all errors and some libraries parse them)

bcoe commented:
@benjamingr do you think a major is necessary, even though someone will opt in with `--experimental-source-maps`, I'm thinking if we drop the experimental we'd probably just switch to `--enable-source-maps` and keep it opt in?

Open to other people's thinking on this.

ljharb commented:
While behind an experimental flag, I’d expect no changes to precipitate a major.

bcoe commented on lib/internal/errors.js:
```diff
@@ -40,9 +40,42 @@ const prepareStackTrace = (globalThis, error, trace) => {
+        if (sourceURL && sourceLineNumber !== undefined &&
+            sourceColumnNumber !== undefined) {
+          annotated +=
+`\n         -> ${sourceURL.replace('file://', '')}:${sourceLineNumber}:${sourceColumnNumber}`;
```
I love the idea too, it was actually @hashseed who brought up the idea that we keep both in the stack-trace; it feels like the best of both worlds.

benjamingr replied:
This will probably break tools that do long stack trace stitching (like bluebird for example) but that's a bridge we can cross later and I also like this idea.

bcoe commented on lib/internal/source_map/source_map.js:
```diff
@@ -0,0 +1,289 @@
+// Copyright 2013 the V8 project authors. All rights reserved.
```
👍 done, I'll probably also publish a copy to npm, once this is a bit more tested, so that it's easier for the community to iterate on (we can upstream any bug fixes to SourceMap.js).

joyeecheung commented:
If it's opt-in via flags I don't think this need to be semver-major. BTW I think it's fine to just use `--enable-source-maps` from the start with an experimental warning.

joyeecheung commented on doc/api/cli.md:
```diff
@@ -184,6 +184,13 @@ added: v11.8.0
+added: v13.x
+-->
+
+Enable experimental Source Map V3 support for stack traces.
```
```suggestion
> Stability: 1 - Experimental

Enable experimental Source Map V3 support for stack traces.
```

joyeecheung commented on lib/internal/errors.js:
```diff
@@ -40,9 +41,45 @@ const prepareStackTrace = (globalThis, error, trace) => {
   }
-  return `${errorString}\n    at ${trace.join('\n    at ')}`;
+
+  if (experimentalSourceMaps === undefined) {
```
Do we really need the `prepareStackTrace` polyfills to work with source maps? I think it would be fine if we choose a different implementation of `prepareStackTrace` during pre-execution and use that instead instead of doing a runtime check for every `error.stack` being accessed. At least we can probably leave those quirks out from the start and see if anyone actually wants them enough to send in a feature request.

bcoe replied:
If I follow, you're suggesting simply swapping in a new `prepareStackTrace` here, if `--enable-source-maps` is set:

https://github.com/nodejs/node/blob/master/lib/internal/bootstrap/node.js#L316

And we'd skip adding:

```js
  // `globalThis` is the global that contains the constructor which
  // created `error`.
  if (typeof globalThis.Error.prepareStackTrace === 'function') {
    return globalThis.Error.prepareStackTrace(error, trace);
  }
```

etc., for the time being, let people ask for it?

joyeecheung replied:
Yes, and we can add a note about this in the doc saying that `--enable-source-maps` does not work with the `Error.prepareStackTrace` hook (I think that would be reasonable)

devsnek replied:
fwiw we still use prepareStackTrace in the repl and in assertions (and maybe when checking the buffer deprecation??)

joyeecheung commented on lib/internal/errors.js:
```diff
@@ -40,9 +41,45 @@ const prepareStackTrace = (globalThis, error, trace) => {
+        const sm = new SourceMap(sourceMap.payload);
+        // Source Map V3 lines/columns use zero-based offsets whereas, in
+        // stack traces, they start at 1/1.
+        const [, , sourceURL, sourceLineNumber, sourceColumnNumber] =
```
nit: I'd just use `url`, `line` and `col` to make the template below shorter.

benjamingr commented:
Sorry for being unclear: I do not think this needs to be major with the flag but I do eventually think Node.js should turn this on by default and _that_ change should be semver-major. 

+1 for Joyee's idea of doing `--enable-source-maps`, I would even take this one step further because I'd like this to be opt-out eventually rather than opt-in and I'd prefer:

`--source-maps=true` which eventually would be the default and users would be able to opt out with `--source-maps=false`

mcollina commented on lib/internal/source_map/source_map_cache.js:
```diff
@@ -0,0 +1,196 @@
+      sourceMap.payload
+    );
+  }
+}
```
This is problematic. Are we freeing it from the cache anytime?

bcoe replied:
what if we put a flag on the entry if it's persisted for the purposes of an exceptional flow, e.g.,

`persitForException`, with an additional parameter, `exceptional: true`, and then when we decorate the stack trace we remove it from cache?

my only concern is the edge-case that an exception happens, and will cause the process to exit, we want that information on disk for source-maps ... will we know by the time that `prepareStackTrace` is called whether the process is slated to be destroyed (@devsnek you might know?).

mcollina replied:
That would be the way to go.

devsnek replied:
you could flag to `prepareStackTrace` whether it was called while formatting a fatal exception, but you still run into a problem:
```js
const e = new Error();
e.stack; // generate stack before fatal exception logic
throw e;
```
not that i think anyone does this *directly*, but it may happen indirectly from custom formatting logic, random console.logs, etc.

ljharb replied:
Given that errors are often wrapped in Promise.reject (ie, not always thrown) it seems pretty likely.

bcoe replied:
I figured out an approach that works for most scenarios; simply move the source-map to being keyed on `error` rather than `module` if compile throws.

bcoe commented:
@LinusU I'm wondering if you have any feedback, as we start to build this out; I imagine you've bumped into some interesting edge cases building [source-map-support](https://github.com/evanw/node-source-map-support).

ljharb commented on lib/internal/source_map/source_map.js:
```diff
@@ -0,0 +1,294 @@
+  /**
+   * @param {SourceMapV3} mappingPayload
+   */
+  _parseMappingPayload: function(mappingPayload) {
```
it seems like using `class` and private class fields would be better here; given that that change actually affects the public API, would it be prudent to make that kind of change now?

(alternatively, all these underscored methods could be kept in closure and `.call`ed on instances)

bcoe replied:
@ljharb this is pretty much a port of `SourceMap.js` from V8, and I'd like to keep it as simple as possible to pull in updates over time, I'd suggest we instead make changes upstream in V8?

bcoe replied:
I guess we'd essentially just replace `_` with `#`? might be a small enough change that it doesn't make importing changes too much of a hassle.

ljharb replied:
you'd have to also switch from old-style prototype construction to a `class`, but yes.

addaleax commented on doc/api/cli.md:
```diff
@@ -184,6 +184,15 @@ added: v11.8.0
 
+### `--source-map-support`
+<!-- YAML
+added: v12.x
```
```suggestion
added: REPLACME
```

bcoe commented:
@benjamingr, @mcollina, @joyeecheung, @devsnek, I believe I've addressed the review comments.

If anyone has the cycles to test on a few projects using source-maps, would love to have a couple more folks kick the tires. I tried to test out a variety of common tools that I know create them.

addaleax commented on doc/api/cli.md:
```diff
@@ -184,6 +184,15 @@ added: v11.8.0
 
 Enable experimental diagnostic report feature.
 
+### `--source-map-support`
```
e anyI’d go with `--source-maps` or `--enable-source-maps`… I don’t hav strong arguments for that, besides maybe that flags usually imply support for some kind of feature already?

bcoe replied:
I like `--enable-source-maps`, personally (I've updated to this and will push soon).

richardlau commented on doc/api/cli.md:
```diff
@@ -184,6 +184,15 @@ added: v11.8.0
 
+### `--source-map-support`
+<!-- YAML
+added: REPLACME
```
```suggestion
added: REPLACEME
```

devsnek commented:
some node internals fail if prepareStackTrace isn't available. I'll open a pr soon to fix this.

addaleax commented:
> some node internals fail if prepareStackTrace isn't available. I'll open a pr soon to fix this.

@devsnek Is that something that should block this PR or is it independent of that?

addaleax commented:
@mcollina As I understand it, this only affects cases where the `.stack` property is accessed and not Error object creation in general.

bcoe commented:
@mcollina, @addaleax is correct, the logic is only applied when stack traces are accessed (not during error creation (I've confirmed this)).

bcoe commented:
@devsnek can this fix land independently from this work?

devsnek commented:
@bcoe 

> can this fix land independently from this work?

some stuff in core will currently fail to work if the flag in this pr is enabled. #29777 moves it to a separate api that both prepareStackTrace callbacks can dispatch safely.

bcoe commented:
@devsnek sounds like perhaps I should wait for #29777? how do you feel about requesting fast track on that PR?

devsnek commented:
@bcoe now that #29777 has landed you can integrate it into your prepareStackTrace callback too.

bcoe commented:
@devsnek is it this logic I should specifically be pulling over?

https://github.com/nodejs/node/pull/29777/files#diff-5a3344c263a73c663dd1cfcb91880fd7R26-R32

devsnek commented:
@bcoe yep! making some sort of helper function to keep it DRY might not go amiss either :)

bcoe commented:
@devsnek I didn't think it was quite worth pulling the logic into a helper; adding the additional export and function felt pretty much equivalent to pulling the code over.

I _did_, however, add a note for future work that it would be nice to support source-maps in the repl (which might add enough logic that we think about refactoring into a helper).

@devsnek, @addaleax, @benjamingr, are you comfortable with signing off on this?

devsnek commented on lib/internal/source_map/source_map_cache.js:
```diff
@@ -136,17 +152,86 @@ function sourceMapCacheToObject() {
+// if possible.
+const prepareStackTrace = (globalThis, error, trace) => {
+  const { SourceMap } = require('internal/source_map/source_map');
+  const ErrorToString = Error.prototype.toString;
```
this line needs to be hoisted out, its purposely saving the original value of `Error.prototype.toString`.

devsnek commented on lib/internal/source_map/source_map_cache.js:
```diff
@@ -136,17 +152,86 @@ function sourceMapCacheToObject() {
+const prepareStackTrace = (globalThis, error, trace) => {
+  const { SourceMap } = require('internal/source_map/source_map');
+  const ErrorToString = Error.prototype.toString;
+  const errorString = ErrorToString.call(error);
```
this line can be moved below the overrideStackTrace check.

benjamingr commented on lib/internal/source_map/source_map.js:
```diff
@@ -0,0 +1,301 @@
+ * @param {SourceMapV3} payload
+ */
+class SourceMap {
+  #reverseMappingsBySourceURL = [];
```
Random note: This is the first time I'm seeing private fields in the Node code base :] I should look more.

bcoe replied:
it was @ljharb's suggestion, might as well use them since we have them available :smile:

ljharb commented on lib/internal/source_map/source_map.js:
```diff
@@ -0,0 +1,301 @@
+             'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
+      base64Map = {};
+      for (let i = 0; i < base64Digits.length; ++i)
+        base64Map[base64Digits.charAt(i)] = i;
```
```suggestion
        base64Map[base64Digits[i]] = i;
```
charAt is needed for maximum back compat, but not needed here.

However, I think you can do `Object.fromEntries(Object.entries(base64Digits))` and replace this entire loop.

bcoe commented:
Landed in e1e2f66
